integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/KnnPredicateSpecificsIT.java

```diff
@@ -0,0 +1,731 @@
+
+			assertThatThrownBy( () -> predicate( f, fieldPath, fieldType ) )
+					.isInstanceOf( SearchException.class )
+					.hasMessageContainingAll(
+							"Cannot use 'predicate:knn' on field '" + fieldPath + "'",
+							"Inconsistent configuration for field '" + fieldPath + "'",
+							"Attribute 'codec' differs",
+							"dimension=4",
+							"dimension=2"
+					);
+		}
+
+		@ParameterizedTest(name = "{0}")
+		@MethodSource("params")
+		void similarity(VectorFieldTypeDescriptor<?> fieldType,
+				SimpleMappedIndex<IndexBinding> index,
+				SimpleMappedIndex<IndexBinding> indexDifferentDimension,
+				SimpleMappedIndex<IndexBinding> indexDifferentBeamWidth,
+				SimpleMappedIndex<IndexBinding> indexDifferentMaxConnection,
+				SimpleMappedIndex<IndexBinding> indexDifferentSimilarity) {
+			SearchPredicateFactory f = index.createScope( indexDifferentSimilarity ).predicate();
+
+			String fieldPath = index.binding().field.get( fieldType ).relativeFieldName;
+
+			assertThatThrownBy( () -> predicate( f, fieldPath, fieldType ) )
+					.isInstanceOf( SearchException.class )
+					.hasMessageContainingAll(
+							"Cannot use 'predicate:knn' on field '" + fieldPath + "'",
+							"Inconsistent configuration for field '" + fieldPath + "'",
+							"Attribute 'codec' differs",
+							"vectorSimilarity=L2",
+							"vectorSimilarity=COSINE"
+					);
+		}
+
+		@ParameterizedTest(name = "{0}")
+		@MethodSource("params")
+		void beamWidth(VectorFieldTypeDescriptor<?> fieldType,
+				SimpleMappedIndex<IndexBinding> index,
+				SimpleMappedIndex<IndexBinding> indexDifferentDimension,
+				SimpleMappedIndex<IndexBinding> indexDifferentBeamWidth,
+				SimpleMappedIndex<IndexBinding> indexDifferentMaxConnection,
+				SimpleMappedIndex<IndexBinding> indexDifferentSimilarity) {
+			StubMappingScope scope = index.createScope( indexDifferentBeamWidth );
+			SearchPredicateFactory f = scope.predicate();
+
+			String fieldPath = index.binding().field.get( fieldType ).relativeFieldName;
+
+			assertThat( scope.query().select()
+					.where( predicate( f, fieldPath, fieldType ).toPredicate() )
+					.fetchAllHits() )
+					.hasSize( 1 );
+		}
+
+		@ParameterizedTest(name = "{0}")
+		@MethodSource("params")
+		void maxConnection(VectorFieldTypeDescriptor<?> fieldType,
+				SimpleMappedIndex<IndexBinding> index,
+				SimpleMappedIndex<IndexBinding> indexDifferentDimension,
+				SimpleMappedIndex<IndexBinding> indexDifferentBeamWidth,
+				SimpleMappedIndex<IndexBinding> indexDifferentMaxConnection,
+				SimpleMappedIndex<IndexBinding> indexDifferentSimilarity) {
+			StubMappingScope scope = index.createScope( indexDifferentMaxConnection );
+			SearchPredicateFactory f = scope.predicate();
+
+			String fieldPath = index.binding().field.get( fieldType ).relativeFieldName;
+
+			assertThat( scope.query().select()
+					.where( predicate( f, fieldPath, fieldType ).toPredicate() )
+					.fetchAllHits() )
+					.hasSize( 1 );
+		}
```

Comment:
Right, at least that's what it seems like. I had these two as such that must match initially, but once I tried it in the test only dimension resulted in an error from Lucene, and as for the similarity -- it just didn't make sense to me to mix them. as for the max connections and beam width, these two are used for storage, I didn't see how it may affect the search algorithm, but at least it's not failing with exceptions 😃

Reply:
Lucene has been notoriously permissive in the past, so just because it doesn't throw an exception, it doesn't mean that it supports what you're doing. They're improving these days, but still...

Considering we don't have many tests that check that the computed distance actually makes sense, I'm not sure we're in a position to promise this will work... 

Honestly I'd just lock all of this down, sending exceptions when the beamwidth/maxConnections are inconsistent, like you do for the dimension. If this bothers someone one day, well... they can open a Jira issue :shrug: